### ISSUE TITLE
Fix initial print dialog window size

### ISSUE THREAD

**Issue by dmik** [bitwiseworks/qtbase-os2#112]
There is a standard print dialog in many example (and real life) apps, for example, `widgets/richtext/textedit`. Currently it opens shrunk so that only the window title bar is visible. You need to resize it manually to be able to configure printing.

My (wild) guess is that this happens because our OS/2 platform plugin doesn't support setting top-level window constraints yet.

**Comment by dmik** [comment]
From what I see during the debug session, this is what happens. First, the dialog is resized according to the size hint. But since it's not shown yet, size/move events are marked as pending (i.e. it's not sent to the platform window that would resize the actual PM window). Then, when the dialog gets eventually shown, `QDialog::showEvent` kicks in and centers the dialog within the main window's geometry. Since this results only in a move event, `QWindow::setPosition` is called instead of `QWindow::setGeometry`. I.e. only the new position is passed, not the new size. Such an order of things causes the platform window to use the old size when moving the window, and ignore its new size completely. This old size then becomes a new (final) size of the window through WM_MOVE/WM_SIZE handling.

In other words, `QWidget`, it's underlying `QWidgetWindow` (a `QWindow` subclass) and its underlying `QPlatformWindow` get out of sync in terms of what the actual window size is. I can't find where it breaks so far. Too many layers to grok it at once. Obviously we don't do something that other platform plugins do in this regard.

**Comment by dmik** [comment]
Ok, I think I found it out. It's a complex issue with multiple points of failure (as usual). The thing is that the initial size of the print dialog is 100 x 30 px (this is somehow the default widget size in Qt). This is a size of the client window area — which makes sense for top-level windows with the system frame. As PM doesn't provide a direct API to get the size of the frame, we use a trick: resize WC_FRAME to the requested size and then see how it positioned its FID_CLIENT in response. Distances from all for FID_CLIENT bounds to the respective WC_FRAME bounds comprise sizes of all four frame margins, respectively, just what we need.

However, for the initial size of 100 x 30 there is a small problem. It's less than the minimal WC_FRAME size. The standard title bar width is 27 px, the standard bottom frame size is 5 px. So, it  gives us a total of 32 px of the *minimum* WC_FRAME height which only includes the frame and no (zero) client area. However, we supplied it only with 30 px. This makes PM go crazy and simply position FID_CLIENT at 0,0. Also, the *actual* WC_FRAME height becomes 31 px instead of the requested 30 px (God knows why). All in all, this leads into a wrong frame margins calculation and a wrong FID_CLIENT position. Due to 1 pixel loss, Qt starts to think that the window is 29 px high.

Note that PM also doesn't allow a width of 100 resizing it to 160. I guess it preserves space for all possible title bar buttons, which is normal per se and should not hurt. So, given the above error, from the Qt POV, the window's client area is 160 x 29 px.

When the window gets resized later (to some bigger size, like 741 x 600 px in case of the print preview dialog), the above mentioned out-of-sync between the Qt size and native window size takes place. Then, when `QDialog` moves the window in an attempt to center it and PM receives the new position but the size is 160 x 29 (because it is out-of-sync at this point). However, due to the frame margin error we got above, Qt uses 31 px as the total frame height instead of 32 px to calculate the proper WC_FRAME height which includes the client area of 29 px. This gives us 29 + 31 = 60 px of the WC_FRAME height. However, since 60 px is greater than the minimum of 32 px, PM works well this time and positions FID_CLIENT correctly reporting its height as 28 px (60 - 32)! Qt picks up this change and thinks the client window size is now 160 x 28 px, i.e. it was resized vertically by 1 px! (While in fact it has been not, not from the Qt point of view, at least). This makes Qt to re-sync the window size and it overwrites the correct, but out-of-sync value of 741 x 600 with 160 x 28 px. Now it's all in sync, but very wrong.

Then a lot of things happens and at some stage Qt checks if the native size gets out of sync with the needed size and forces a native resize if so. However, on OS/2, since it has been already synced, nothing happens. The window remains at its wrong size.

On macOS (and other platforms), the native window size will remain 100 x 30 — no 1 px error and no resize event that wolud sync it. So, as a result, the out-of-sync check code kicks in and resizes the window to the proper size of 741 x 600 px. Finita la comedia.

PS. It only took around 20 min for me to write it all down. You can imagine how long it took to **find** it all out. But now I know how to fix it, definitely.
